### PR TITLE
Vulkan logical device setup and disposing of vulkan objects

### PR DESCRIPTION
**Pull request by Carlbaum**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -30,6 +30,7 @@ add_library(direwolf
     src/renderengine.cpp
     include/direwolf/renderengine.h
     # TODO: this shouldn't be here.. remove when vulkan stuff is included in the direwolf wrappers
+    include/direwolf/vulkan/vulkancommons.h
```
Comment (novalain) [inline CMakeLists.txt:33]: NIT: vulkancommon or vulkanutils imo

File: examples/example_vulkan.cpp [hunk 2]
```diff
@@ -45,9 +47,177 @@ void testNonEngineIntergratedVulkan()
 
     std::vector<VkPhysicalDevice> physicalDevices = GetPhysicalDevices(vulkanInstance);
 
+    VkPhysicalDevice selectedDevice = VK_NULL_HANDLE;
+    std::vector<QueueInfo> queueInfos; // TODO: should be a set?
+
+    // TODO: Only temp.. These are all the extensions available on Dell XPS with GTX 1050.. using a subset of them as desiredExtensions now
+    const std::vector<char const *> desiredDeviceExtensions{
```
Comment (novalain) [inline examples/example_vulkan.cpp:54]: Are you supposed to know about what extensions your device has?

File: examples/example_vulkan.cpp [hunk 3]
```diff
@@ -45,9 +47,177 @@ void testNonEngineIntergratedVulkan()
...
+        ci.pNext = nullptr;
+        ci.flags = 0;
+        ci.queueFamilyIndex = info.familyIndex;
+        ci.queueCount = static_cast<uint32_t>(info.priorities.size());
+        ci.pQueuePriorities = info.priorities.size() > 0 ? info.priorities.data() : nullptr;
+        queueCreateInfos.push_back(ci);
     }
+
+    // Logical device
+    // TODO: Should be done in initializer list.. but better for clarity now
```
Comment (novalain) [inline examples/example_vulkan.cpp:191]: Nah I think this is fine

File: include/direwolf/vulkan/vulkancommons.h [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+#include <string>
+
+#if defined (_WIN32)
+#include <windows.h>
+#endif // _WIN32
+
+namespace dw::vulkan {
+
+#if defined (_WIN32)
+typedef HMODULE VulkanRTLPtr;
```
Comment (novalain) [inline include/direwolf/vulkan/vulkancommons.h:15]: nit: 2 spaces indent?

File: include/direwolf/vulkan/vulkancommons.h [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+#elif defined (__linux) || defined (__APPLE__)
+typedef void* VulkanRTLPtr;
+#else
+typedef void* VulkanRTLPtr;
```
Comment (novalain) [inline include/direwolf/vulkan/vulkancommons.h:19]: what this case? should never happen?

File: examples/example_vulkan.cpp [hunk 6]
```diff
@@ -45,9 +47,177 @@ void testNonEngineIntergratedVulkan()
...
+    VkDeviceCreateInfo deviceCreateInfo;
+    deviceCreateInfo.sType = VK_STRUCTURE_TYPE_DEVICE_CREATE_INFO;
+    deviceCreateInfo.pNext = nullptr;
+    deviceCreateInfo.flags = 0;
+    deviceCreateInfo.queueCreateInfoCount = static_cast<uint32_t>(queueCreateInfos.size());
+    deviceCreateInfo.pQueueCreateInfos = queueCreateInfos.size() > 0 ? queueCreateInfos.data() : nullptr;
+    deviceCreateInfo.enabledLayerCount = 0;
+    deviceCreateInfo.ppEnabledLayerNames = nullptr;
+    deviceCreateInfo.enabledExtensionCount = static_cast<uint32_t>(desiredDeviceExtensions.size());
+    deviceCreateInfo.ppEnabledExtensionNames = desiredDeviceExtensions.size() > 0 ? &desiredDeviceExtensions[0] : nullptr;
```
Comment (novalain) [inline examples/example_vulkan.cpp:201]: can we pass all of this stuff as some metadata/config later?

File: examples/example_vulkan.cpp [hunk 7]
```diff
@@ -45,9 +47,177 @@ void testNonEngineIntergratedVulkan()
...
+
+    VkDevice logicalDevice = CreateLogicalDevice(selectedDevice, deviceCreateInfo);
+    if (logicalDevice == VK_NULL_HANDLE) {
+        std::cerr << "ERROR: Failed to create logical device" << std::endl;
+        return;
+    }
+    ////////////////////////////////////////////////////////////////////////////////////////////////////
+    // When everything is done, we should close down everything properly.
+    // This should be done in reverse order
+    std::cout << "\n\nEverything succeeded!\n\nDestroying vulkan objects...\n" << std::endl;
```
Comment (novalain) [inline examples/example_vulkan.cpp:215]: we should create a common logger soon

File: src/vulkan/vulkancommons.cpp [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+#include <iomanip>
+#include <sstream>
+
+#include "vulkanfunctions.h"
+
+namespace dw::vulkan {
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+// overloaded operator<<
```
Comment (novalain) [inline src/vulkan/vulkancommons.cpp:11]: redundant comment remove

File: src/vulkan/vulkansetup.cpp [hunk 9]
```diff
@@ -185,4 +185,93 @@ bool IsExtensionSupported(const char* extension, const std::vector<VkExtensionPr
...
+    }
+
+    return queueFamilyProperties;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+bool IsQueueFamilySupportingFlags(const VkQueueFamilyProperties& queueFamily, const VkQueueFlags flagsToCheck)
+{
+    if (queueFamily.queueCount > 0 && (queueFamily.queueFlags & flagsToCheck) == flagsToCheck) {
```
Comment (novalain) [inline src/vulkan/vulkansetup.cpp:212]: imo `return queueFamily.queueCount > 0 && (queueFamily.queueFlags & flagsToCheck) == flagsToCheck`;

File: src/vulkan/vulkansetup.cpp [hunk 10]
```diff
@@ -185,4 +185,93 @@ bool IsExtensionSupported(const char* extension, const std::vector<VkExtensionPr
...
+
+VkDevice CreateLogicalDevice(const VkPhysicalDevice& physicalDevice, const VkDeviceCreateInfo& deviceCreateInfo)
+{
+    VkDevice logicalDevice;
+    if (vkCreateDevice(physicalDevice, &deviceCreateInfo, nullptr, &logicalDevice) != VK_SUCCESS || logicalDevice == VK_NULL_HANDLE) {
+        std::cerr << "ERROR: Failed to create logical device" << std::endl;
+        return VK_NULL_HANDLE;
+    }
+
+    // TODO: might not be nice to do this here........... move elsewhere?
```
Comment (novalain) [inline src/vulkan/vulkansetup.cpp:242]: why not

**Review by novalain — APPROVED** [review]
Looks good! Only nit picks

File: CMakeLists.txt (hunk 1 above)
Comment (Carlbaum) [reply CMakeLists.txt:33]: Well.. they have different purposes, but agreed. Vulkan Utils should probably be removed.
The way it is now, commons should be inclued in a header, utils in a source file

File: examples/example_vulkan.cpp (hunk 2 above)
Comment (Carlbaum) [reply examples/example_vulkan.cpp:54]: Yup.. you're supposed to provide the extensions you'll need. But this we can hide within DireWolf

File: include/direwolf/vulkan/vulkancommons.h (hunk 5 above)
Comment (Carlbaum) [reply include/direwolf/vulkan/vulkancommons.h:19]: I wanted it in there just for modularity, but sure, I'll remove it

File: examples/example_vulkan.cpp (hunk 6 above)
Comment (Carlbaum) [reply examples/example_vulkan.cpp:201]: Yeah, something like that

File: src/vulkan/vulkansetup.cpp (hunk 9 above)
Comment (Carlbaum) [reply src/vulkan/vulkansetup.cpp:212]: hahahah.. yeah ofc :P
My bääd!

**Comment by Carlbaum** [comment]
I fucked up, hadn't staged the changes ;)
Can I revert the commit or amend it somehow?
If I do revert on the PR, won't that add the revert as an extra commit?

**Comment by novalain** [comment]
Either is fine.
1. Open revert PR and put it into master, then merge fixed PR again (2 extra commits, 3 in total)
2. Rip commit from master (force push), then merge fixed PR again (no extra commits)

**Pull request merged**